[RLlib; Docs] Auto API reference pages overhaul: `rllib/policy` and `rllib/agents` packages.

sven1977 opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

Auto API reference pages overhaul: `rllib/policy` and `rllib/agents` packages.

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

<!-- Please give a short summary of the change and the problem this solves. -->

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

richardliaw commented on doc/source/rllib/package_ref/policy.rst:
```diff
@@ -0,0 +1,49 @@
+The ``Policy`` class contains functionality to compute actions for decision making
+in an environment, as well as computing loss(es) and gradients, updating a neural
+network model as well as postprocessing a collected environment trajectory.
+One or more ``Policy`` objects sit inside a `RolloutWorker's`` "policy_map" and
```
We should make sure that `RolloutWorker` is linked to the actual docstring, right?

richardliaw replied:
Is RolloutWorker a concept that we want to expose to the user?

sven1977 replied:
Yes, I would say that the RolloutWorker API is definitely something the user should be able to sub-class and use in custom Trainers. Both ES and ARS use custom workers.

sven1977 replied:
done

richardliaw commented on doc/source/rllib/package_ref/policy.rst:
```diff
@@ -0,0 +1,49 @@
+    needs.
+
+
+Building Custom Policy Classes
+++++++++++++++++++++++++++++++
+
+.. warning::
+    It is not recommended anymore to use the ``build_policy_class()`` or ``build_tf_policy()``
+    utility functions for creating custom Policy sub-classes.
+    Instead, follow the simple guidelines here for directly sub-classing from
+    either one of the built-in types.
+
+In order to create a custom Policy, simply sub-class ``Policy`` (for a generic,
+framework-agnostic policy), ``TorchPolicy`` (for a PyTorch specific policy), or
+``TFPolicy`` (for a TensorFlow specific policy) and override one or more of their
+methods. Those are in particular:
+
+* compute_actions_from_input_dict
+* postprocess_trajectory
+* loss
```
I think 'customization' should be in a separate page, or at least -- the user should be exposed to the default policies before seeing the custom instructions.

sven1977 replied:
doe

richardliaw commented on doc/source/rllib/package_ref/policy.rst:
```diff
@@ -0,0 +1,49 @@
+.. https://docs.google.com/drawings/d/1eFAVV1aU47xliR5XtGqzQcdvuYs2zlVj1Gb8Gg0gvnc/edit
+.. figure:: ../../images/rllib/policy_classes_overview.svg
+    :align: left
+
+    **RLlib's Policy class hierarchy:** Policies are deep-learning framework
+    specific as they hold functionality to handle a computation graph (e.g. a
+    TensorFlow 1.x graph in a session). You can define custom policy behavior
+    by sub-classing either of the available, built-in classes, depending on your
+    needs.
+
```
If I'm on this page, I want to be able to click into these specific policies (and have a summary of the different policies here)

sven1977 replied:
You do have the overview through the toc above, no?

richardliaw commented on doc/source/rllib/package_ref/policy.rst:
```diff
@@ -0,0 +1,49 @@
+``TFPolicy`` (for a TensorFlow specific policy) and override one or more of their
+methods. Those are in particular:
+
+* compute_actions_from_input_dict
+* postprocess_trajectory
```
You should link to the implementation or examples?

richardliaw commented on doc/source/rllib/package_ref/trainer.rst:
```diff
@@ -0,0 +1,50 @@
+Building Custom Trainer Classes
++++++++++++++++++++++++++++++++
+
+.. warning::
+    It is not recommended anymore to use the `build_trainer()` utility function
+    for creating custom Trainer sub-classes.
```
This is old right? Can you say "As of X version of Ray, it is not recommended..."

sven1977 replied:
done

richardliaw commented on doc/source/rllib/package_ref/trainer.rst:
```diff
@@ -0,0 +1,50 @@
+    :align: left
+
+    **A typical RLlib Trainer object:** The components sitting inside a Trainer are
+    normally: N ``WorkerSets`` (each consisting of one local ``RolloutWorker`` and zero or
+    more @ray.remote ``RolloutWorkers``),
```
Can you link to these concepts? Or should you even introduce them here so early on?

I'm thinking that this seems a bit more like implementation. If a user is trying to understand how to use their Trainer, and they land on this page first, it will be quite confusing

sven1977 replied:
done

sven1977 replied:
I think introducing this here is very useful. There is often a confusion as to "Policy" vs "Trainer" vs "Worker", etc..

We do have the new svg here showing a few high-level components that go into a Trainer. I think this explains quite well, how RLlib handles distributed learning and what the "Policy"/"Trainer"/"Worker" concepts mean.

sven1977 replied:
Lined all classes mentioned to their code.

sven1977 commented:
@richardliaw , please take another look. Addressed all your questions.

richardliaw commented on doc/source/rllib/package_ref/policy.rst:
```diff
@@ -0,0 +1,32 @@
+in an environment, as well as computing loss(es) and gradients, updating a neural
+network model as well as postprocessing a collected environment trajectory.
+One or more ``Policy`` objects sit inside a
+`RolloutWorker's <https://github.com/ray-project/ray/blob/master/rllib/evaluation/rollout_worker.py>`_
```
If we expect the user to know the rollout worker, can you make sure there is a reference to the API?

sven1977 replied:
ok, done

richardliaw commented on doc/source/rllib/package_ref/policy.rst:
```diff
@@ -0,0 +1,32 @@
+in an environment, as well as computing loss(es) and gradients, updating a neural
+network model as well as postprocessing a collected environment trajectory.
+One or more ``Policy`` objects sit inside a
+`RolloutWorker's <https://docs.ray.io/en/master/rllib-package-ref.html#module-ray.rllib.evaluation>`_
```
1. Shouldn't this link to RolloutWorker the object in particular?
2. We should not have *any* references to master. All references need to link to the current documentation, or else these links easily get stale.

richardliaw replied:
(see comment on PR)

sven1977 replied:
done

richardliaw reviewed (changes requested):
Sven, we should not have any references to the github or documentation master. 

All references need to link to the current documentation, or else these links easily get stale.

If you need to show code, you should use a relative reference. Let me know if you have questions.

richardliaw commented:
```diff
diff --git a/doc/source/rllib/package_ref/policy.rst b/doc/source/rllib/package_ref/policy.rst
index 1b24fcb5e..7e802afa8 100644
--- a/doc/source/rllib/package_ref/policy.rst
+++ b/doc/source/rllib/package_ref/policy.rst
@@ -15,8 +15,7 @@ Policy APIs
 The ``Policy`` class contains functionality to compute actions for decision making
 in an environment, as well as computing loss(es) and gradients, updating a neural
 network model as well as postprocessing a collected environment trajectory.
-One or more ``Policy`` objects sit inside a
-`RolloutWorker's <https://docs.ray.io/en/master/rllib-package-ref.html#module-ray.rllib.evaluation>`_
+One or more ``Policy`` objects sit inside a :py:class:`~ray.rllib.evaluation.RolloutWorker`'s
 `policy_map <https://github.com/ray-project/ray/blob/master/rllib/policy/policy_map.py>`_ and
 are - if more than one - are selected based on a multi-agent ``policy_mapping_fn``,
 which maps agent IDs to a policy ID.
```


This will allow the user to directly identify the docstring to read (whereas it currently links to something in master, and not even the `RolloutWorker` individually).

You'll need to also make sure `policy_map` shows up in the references, since right now you're linking to code.

richardliaw commented on doc/source/rllib/package_ref/policy.rst:
```diff
@@ -0,0 +1,32 @@
+Policy APIs
+===========
+
+.. toctree::
+   :maxdepth: 1
+
+   policy/policy.rst
+   policy/tf_policies.rst
+   policy/torch_policy.rst
+   policy/custom_policies.rst
```
unfortunately, after using this page a bit, I've realized that this is a bit hard to realize that it is linking to actual API references. Maybe you can move this toctree to the bottom of the page with a section called 'API Reference'?

sven1977 replied:
done

sven1977 commented:
@richardliaw 
1) Changed all links to go directly to the docstrings.
2) No more links to master code (classes)
3) Moved the 2 new toctrees for policy and env to the bottom of the file under a new sub-header: `Env|Policy API Reference`

richardliaw commented:
nice!!

On Wed, Nov 24, 2021 at 3:27 AM Sven Mika ***@***.***> wrote:

> @richardliaw <https://github.com/richardliaw>
>
>    1. Changed all links to go directly to the docstrings.
>    2. No more links to master code (classes)
>    3. Moved the 2 new toctrees for policy and env to the bottom of the
>    file under a new sub-header: Env|Policy API Reference
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ray-project/ray/pull/20537#issuecomment-977787581>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABCRZZO7V2EEKY4I4S5RGWLUNTDYFANCNFSM5IJWOD3A>
> .
>

richardliaw commented on doc/source/rllib/package_ref/policy/custom_policies.rst:
```diff
@@ -0,0 +1,26 @@
+* :py:meth:`~ray.rllib.policy.policy.Policy.postprocess_trajectory`
+* :py:meth:`~ray.rllib.policy.policy.Policy.loss`
+
+`See here for an example on how to override TorchPolicy <https://github.com/ray-project/ray/blob/master/rllib/agents/ppo/ppo_torch_policy.py>`_.
```
in the future, lets make sure we version control these files. Tune has a couple examples here.

richardliaw replied:
you basically create an orphan file and link to it. see the tune examples page.

richardliaw commented on doc/source/rllib/package_ref/trainer.rst:
```diff
@@ -0,0 +1,57 @@
+* :py:meth:`~ray.rllib.agents.trainer.Trainer.step_attempt`
+* :py:meth:`~ray.rllib.agents.trainer.Trainer.execution_plan`
+
+`See here for an example on how to override Trainer <https://github.com/ray-project/ray/blob/master/rllib/agents/pg/pg.py>`_.
```
same here

richardliaw reviewed (approved):
nice! much better
